src/pycea/tl/permutation.py

```diff
@@ -0,0 +1,270 @@
+    data: pd.DataFrame,
+    n_permutations: int,
+    reduction_fn: Callable[[np.ndarray], Union[np.ndarray, float]],
+    difference_fn: Callable[[Union[np.ndarray, float], Union[np.ndarray, float]], float],
+    n_right: int,
+    n_left: int
+) -> np.ndarray:
+    """
+    Randomly permute row assignments across two groups and record
+    (right_stat - left_stat) for each permutation.
+
+    Parameters
+    ----------
+    data : pd.DataFrame
+        Full dataset to split each permutation.
+    n_permutations : int
+        Number of permutations to run.
+    reduction_fn
+        Callable function that can reduce the data from all the leaves of a given split to a vector or scalar.
+    difference_fn
+        Callable function that takes to outputs from reduction_fn (one from each side of a node) and returns a scalar.
+    n_right : int
+        Size of the "right" group in each permutation.
+    n_left : int
+        Size of the "left" group in each permutation.
+
+    Returns
+    -------
+    np.ndarray
+        Array of length n_permutations with permutation statistics.
+    """
+    n = len(data)
+
+    permutation_vals = np.zeros(n_permutations, dtype=float)
+
+    for i in range(n_permutations):
+        # Randomly permute the row indices
+        perm = np.random.permutation(n)
+
+        # Take the first n_left as left, next n_right as right
+        left_idx = perm[:n_left]
+        right_idx = perm[n_left:n_left + n_right]
+
+        left_df = data.iloc[left_idx]
+        right_df = data.iloc[right_idx]
+        left_stat = reduction_fn(left_df.to_numpy())
+        right_stat = reduction_fn(right_df.to_numpy())
+
+        permutation_vals[i] = difference_fn(left_stat, right_stat)
+
+    return permutation_vals
+
+def difference(a: float, b: float) -> float:
+    return a - b
+
+@overload
+def split_permutation_test(
+    tdata: td.TreeData,
+    keys: str | Sequence[str],
+    reduction_fn: Callable[[np.ndarray], np.ndarray],
+    difference_fn: Callable[[np.ndarray, np.ndarray], float],
+    permutation_test: Literal[True, False] = True,
+    n_permutations: int = 1000,
+    min_required_permutations: int = 40,
+    keys_added: str | Sequence[str] | None = None,
+    tree: str | Sequence[str] | None = None,
+    copy: Literal[True, False] = True
+) -> pd.DataFrame: ...
+@overload
+def split_permutation_test(
+    tdata: td.TreeData,
+    keys: str | Sequence[str],
+    reduction_fn: Callable[[np.ndarray], np.ndarray],
+    difference_fn: Callable[[np.ndarray, np.ndarray], float],
+    permutation_test: Literal[True, False] = True,
+    n_permutations: int = 1000,
+    min_required_permutations: int = 40,
+    keys_added: str | Sequence[str] | None = None,
+    tree: str | Sequence[str] | None = None,
+    copy: Literal[True, False] = False
+) -> None: ...
+def split_permutation_test(
+    tdata: td.TreeData,
+    keys: str | Sequence[str],
+    reduction_fn: Callable[[np.ndarray], Union[np.ndarray, float]] = np.mean,
```

Comment:
Currently when `split_permutatation_test` is run on an obsm key like PCA, `reduction_fn` sums the PCs to a scalar before `difference_fn` calculates the distance. I think it would make more sense to average each PC separately and then calculate a distance between the left and right PCA vectors. Maybe rename`reduction_fn` as `aggregate` and `difference_fn` as `metric`. `metric` could be any sklearn distance function similar to `pycea.tl.distance`. These arguments should be able to accept string inputs for common options: `metric = 'euclidean'`